virttest/migration_template.py

```diff
@@ -0,0 +1,563 @@
+        self.uptime = {}
+        for vm in self.migrate_vms:
+            self.uptime[vm.name] = vm.uptime(connect_uri=vm.connect_uri)
+
+        if self.params['migrate_source_host_cn']:
+            src_uri = get_uri(self.params['uri_type'],
+                              self.params['remote_protocol'],
+                              self.params['migrate_source_host_cn'])
+        else:
+            src_uri = get_uri(self.params['uri_type'],
+                              self.params['remote_protocol'],
+                              self.params['migrate_source_host'])
+
+        # Migrate vm back to src host
+        logging.debug("Start to migrate vm back to src host")
+        self.obj_migration.do_migration(self.migrate_vms, self.dest_uri,
+                                        src_uri, "orderly",
+                                        options=self.params['virsh_migrate_options'],
+                                        thread_timeout=self.params['migrate_thread_timeout'],
+                                        ignore_status=False,
+                                        virsh_uri=self.dest_uri,
+                                        shell=True)
+
+        # Check migration result: succeed or fail with expected error
+        logging.debug("Check migration result: whether succeed"
+                      "or fail with expected error")
+        self.obj_migration.check_result(self.obj_migration.ret, self.params)
+
+        # Set vm connect_uri to src_uri after migration back
+        for vm in self.migrate_vms:
+            vm.connect_uri = self.src_uri
+
+        logging.debug("Do post migration check after migrate back to src")
+        self.obj_migration.post_migration_check(self.migrate_vms, self.params,
+                                                self.uptime)
+        # Restore vm.connect_uri as it is changed in post_migration_check()
+        for vm in self.migrate_vms:
+            vm.connect_uri = self.src_uri
+
+    def _post_migrate_back(self):
+        """
+        Operations after migration, e.g.
+        Check domain xml, check device in vm, etc
+        """
+        raise NotImplementedError
+
+    def _set_vm_shared_disk(self):
+        """
+        Set shared disk in vm xml, only support nfs storage for now:
+        1) change the source of the first disk of vm to shared disk
+        2) change the disk cache mode to none
+        """
+        logging.debug("Prepare shared disk in vm xml for live migration")
+        if self.params['storage_type'] == 'nfs':
+            logging.debug("Prepare nfs shared disk in vm xml")
+            for vm in self.migrate_vms:
+                libvirt.update_vm_disk_source(vm.name, self.params['nfs_mount_dir'])
+                libvirt.update_vm_disk_driver_cache(vm.name, driver_cache="none")
+        else:
+            # TODO:other storage types
+            self.test.cancel("Other storage type is not supported for now")
```

Comment:
“Only nfs storage type is supported so far”
Using positive words seems better.

Reply:
OKey